### ISSUE TITLE
SSH Passphrase and Forward Agent

### ISSUE THREAD

**Issue by Skibooss** [aenima-x/BlackHole#12]
Hello,

Apparently BlackHole does not support DSA or RSA key with a passphrase. Do you think it is possible to implement ?
I think the big security problem is the private key in the database. Do you think this can be get around by used the ForwardAgent in our ssh config and BlackHole get the key this way ?

Thanks

**Comment by aenima-x** [comment]
Yes, I get what you mean. 
But the idea behind BlackHole is to act as a layer between the users and the targets.
They know nothing about authentication, and by this all can be dynamic. 
If you change the users profile It now have a new different set of targets.
Take in consideration that this is not designed as a pesonal session manager, but for a big company as a security and audit tool.
You are right on the security issue of having the keys in a database, that's why my advice is to disable scp and sftp so nobody can acces to the django settings file to get the user and password. Even better if the database is protected on an external server.
I have never used Agent Forwarding, so  if you think theres a way that it could be compatible with it please let me know and I will think about it.

**Comment by Skibooss** [comment]
Yes indeed, I try to implement this on the server of my company for security issue. Everybody who connect to the server have his own private & public key and the most of them have passphrase. Maybe the passphrase can be ask to the user during the connection.

For the forward agent, I read that paramiko implement that (request_forward_agent()). It can be a good solution to avoid the private key in the database.

**Comment by aenima-x** [comment]
Yes but like I said, the users must know the passphrase.
That that is something I don't want. I know it's more secure, but is not the functionality I was looking for.

**Comment by aenima-x** [comment]
Another option to be more "secure" could be encrypt de keys in the database

**Comment by BenoitNorrin** [comment]
+1 for the encrytion in database

**Comment by Skibooss** [comment]
Yes +1 for database encryption.

Can I try to develop an option for ForwardAgent or can I fork it ? :octocat: 
Tanks !

**Comment by aenima-x** [comment]
of course go ahead, theres a cool django app called encrypted fields. I will use it to do the databse encription

**Comment by Skibooss** [comment]
Oh sorry, I talk about an option for Forward the private key trough BlackHole (pramiko SSH Forward Agent) not the database encription. Actually, if it works I won't need the private key in the database.

**Comment by aenima-x** [comment]
Yes I understood, it was just a comment.
